drivers/sensor/tmag5273/tmag5273.c

```diff
@@ -0,0 +1,1097 @@
+
+	switch (drv_cfg->averaging) {
+	case TMAG5273_DT_AVERAGING_2X:
+		val = TMAG5273_CONV_AVG_2;
+		break;
+	case TMAG5273_DT_AVERAGING_4X:
+		val = TMAG5273_CONV_AVG_4;
+		break;
+	case TMAG5273_DT_AVERAGING_8X:
+		val = TMAG5273_CONV_AVG_8;
+		break;
+	case TMAG5273_DT_AVERAGING_16X:
+		val = TMAG5273_CONV_AVG_16;
+		break;
+	case TMAG5273_DT_AVERAGING_32X:
+		val = TMAG5273_CONV_AVG_32;
+		break;
+	case TMAG5273_DT_AVERAGING_NONE:
+		__fallthrough;
+	default:
+		val = TMAG5273_CONV_AVG_1;
+		break;
+	}
+
+	drv_data->conversion_time_us = TMAG5273_T_CONVERSION_US(
+		(val), (drv_cfg->x_axis + drv_cfg->y_axis + drv_cfg->z_axis));
+
+	regdata |= FIELD_PREP(TMAG5273_CONV_AVG_MSK, val);
+	regdata |= FIELD_PREP(TMAG5273_I2C_READ_MSK, I2C_READ_MODE_STANDARD);
+
+	retval = i2c_reg_write_byte_dt(&drv_cfg->i2c, TMAG5273_REG_DEVICE_CONFIG_1, regdata);
+	if (retval < 0) {
+		LOG_ERR("error setting DEVICE_CONFIG_1");
+		return -EIO;
+	}
+
+	/* REG_DEVICE_CONFIG_2 */
+	regdata = 0;
+
+	val = (drv_cfg->low_noise_mode) ? TMAG5273_LOWNOISE_MODE : TMAG5273_LOWPOWER_MODE;
+	regdata |= FIELD_PREP(TMAG5273_LP_LN_MSK, val);
+
+	regdata |= FIELD_PREP(TMAG5273_I2C_GLITCH_FILTER_MSK, TMAG5273_I2C_GLITCH_FILTER_ON);
+
+	val = (drv_cfg->trigger_conv_via_int) ? TMAG5273_TRIGGER_MODE_INT
+					      : TMAG5273_TRIGGER_MODE_I2C;
+	regdata |= FIELD_PREP(TMAG5273_TRIGGER_MODE_MSK, val);
+
+	val = (drv_cfg->operation_mode == TMAG5273_DT_OPER_MODE_CONTINUOUS)
+		      ? TMAG5273_OPERATING_MODE_CONTINUOUS
+		      : TMAG5273_OPERATING_MODE_STANDBY;
+	regdata |= FIELD_PREP(TMAG5273_OPERATING_MODE_MSK, val);
+
+	retval = i2c_reg_write_byte_dt(&drv_cfg->i2c, TMAG5273_REG_DEVICE_CONFIG_2, regdata);
+	if (retval < 0) {
+		LOG_ERR("error setting DEVICE_CONFIG_2");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief sets the \c SENSOR_CONFIG_1 and \c SENSOR_CONFIG_2 registers
+ *
+ * @param drv_cfg configuration of the TMAG5273 instance
+ *
+ * @retval 0 if everything was okay
+ * @retval -EIO on communication errors
+ */
+static inline int tmag5273_init_sensor_settings(const struct tmag5273_config *drv_cfg)
+{
+	int retval;
+	uint8_t regdata;
+
+	/* REG_SENSOR_CONFIG_1 */
+	regdata = 0;
+	uint8_t val = ((drv_cfg->x_axis) ? TMAG5273_MAGNETIC_CH_X_EN : TMAG5273_MAGNETIC_CH_NONE) |
+		      ((drv_cfg->y_axis) ? TMAG5273_MAGNETIC_CH_Y_EN : TMAG5273_MAGNETIC_CH_NONE) |
+		      ((drv_cfg->z_axis) ? TMAG5273_MAGNETIC_CH_Z_EN : TMAG5273_MAGNETIC_CH_NONE);
+
+	regdata |= FIELD_PREP(TMAG5273_MAG_CH_EN_MSK, val);
+
+	retval = i2c_reg_write_byte_dt(&drv_cfg->i2c, TMAG5273_REG_SENSOR_CONFIG_1, regdata);
+	if (retval < 0) {
+		LOG_ERR("error setting SENSOR_CONFIG_1");
+		return -EIO;
+	}
+
+	/* REG_SENSOR_CONFIG_2 */
+	regdata = 0;
+
+	val = (drv_cfg->ch_mag_gain_correction == TMAG5273_DT_CORRECTION_CH_2)
+		      ? TMAG5273_MAG_GAIN_CORRECTION_CH_2
+		      : TMAG5273_MAG_GAIN_CORRECTION_CH_1;
+
+	regdata |= FIELD_PREP(TMAG5273_GAIN_CORRECTION_CH_MSK, val);
```

Comment:
same ere, would not bother with FIELD_PREP, at least I don't think it makes the code any easier to understand

```c
if (drv_cfg->ch_mag_gain_correction == TMAG5273_DT_CORRECTION_CH_2) {
        regdata |= BIT(TMAG5273_GAIN_CORRECTION_CH2);
}
```

applies to a bunch of other single bit fields, don't quite see the point of using FIELD_PREP for everything, and that's not how most drivers are written. Sorry I understand the pattern is coming from a different coding style, just looking for coherency withing the code base.

Reply:
I will go over it. Tbh, the code base is not really coherent here. There are (custom) macros setting stuff, functions setting stuff, stuff which is written directly etc. (at least in the sensor folder). Again, some more rules here would really help.

I will change it nonetheless to increase readability.

Reply:
Yeah there's more than few cases where IMO we should have been more rigorous about sticking to simpler patterns and the fact that the project accepted HALs for sensors does not help, but I think we should try and keep new drivers polished at least. Final go/no-go is up to the subsystem maintainer, I'm just pointing out pattern I spot commonly.

Reply:
I've done a step in between, next days I may not come to do sth. here. If it's good enough then resolve, otherwise I will go to the `BIT` like definition in the next step.

Reply:
I restructured the setting parts and changed the value definitions so that `FIELD_GET` etc. shouldn't only be necessary in special cases.

I keep this conversation open since I have no hardware to test until beginning of January, to avoid a (maybe) untested merge.